File: Packages/CineleafCore/Sources/CineleafCore/ProjectValidation.swift
import Foundation

public enum ProjectValidationError: Error, Equatable, Sendable {
    case unsupportedFormatVersion(Int)
    case invalidCanvas
    case duplicateIdentifier(UUID)
    case invalidTime(UUID)
    case incompatibleTrack(UUID)
    case missingAsset(UUID)
    case missingText(UUID)
    case overlappingClips(UUID, UUID)
    case invalidOpacity(UUID)
    case invalidAudioVolume(UUID)
    case invalidTransform(UUID)
    case invalidFade(UUID)
    case invalidPlaybackRate(UUID)
    case invalidColorAdjustments(UUID)
    case invalidEffect(UUID)
    case invalidTransition(UUID)
    case invalidKeyframe(UUID)
    case invalidRole(UUID)
}

public enum ProjectValidator {
    public static func validate(_ project: CineleafProject) throws {
        guard project.formatVersion == CineleafProject.currentFormatVersion else {
            throw ProjectValidationError.unsupportedFormatVersion(project.formatVersion)
        }
        guard project.canvas.width > 0 && project.canvas.height > 0 else {
            throw ProjectValidationError.invalidCanvas
        }

        var identifiers = Set<UUID>()
        guard identifiers.insert(project.id).inserted else {
            throw ProjectValidationError.duplicateIdentifier(project.id)
        }
        let assetIDs = Set(project.assets.map(\.id))
        let assetsByID = Dictionary(uniqueKeysWithValues: project.assets.map { ($0.id, $0) })
        for asset in project.assets where !identifiers.insert(asset.id).inserted {
            throw ProjectValidationError.duplicateIdentifier(asset.id)
        }
        for marker in project.timeline.markers {
            guard identifiers.insert(marker.id).inserted else {
                throw ProjectValidationError.duplicateIdentifier(marker.id)
            }
            guard marker.time >= .zero else { throw ProjectValidationError.invalidTime(marker.id) }
        }

        for track in project.timeline.tracks {
            guard identifiers.insert(track.id).inserted else {
                throw ProjectValidationError.duplicateIdentifier(track.id)
            }

            let ordered = track.clips.sorted { $0.timelineStart < $1.timelineStart }
            for (index, clip) in ordered.enumerated() {
                guard identifiers.insert(clip.id).inserted else {
                    throw ProjectValidationError.duplicateIdentifier(clip.id)
                }
                guard clip.timelineStart >= .zero, clip.sourceStart >= .zero, clip.duration > .zero else {
                    throw ProjectValidationError.invalidTime(clip.id)
                }
                guard clip.playbackRate.isFinite, (0.25...4).contains(clip.playbackRate) else {
                    throw ProjectValidationError.invalidPlaybackRate(clip.id)
                }
                guard clip.kind.compatibleTrack == track.kind else {
                    throw ProjectValidationError.incompatibleTrack(clip.id)
                }
                if clip.kind == .text {
                    guard clip.textStyle != nil, clip.assetID == nil else {
                        throw ProjectValidationError.missingText(clip.id)
                    }
                } else if let assetID = clip.assetID {
                    guard assetIDs.contains(assetID) else {
                        throw ProjectValidationError.missingAsset(assetID)
                    }
                    if clip.kind != .image,
                       let sourceDuration = assetsByID[assetID]?.metadata.duration,
                       clip.sourceStart + scaled(clip.duration, by: clip.playbackRate) > sourceDuration {
                        throw ProjectValidationError.invalidTime(clip.id)
                    }
                } else {
                    throw ProjectValidationError.missingAsset(clip.id)
                }
                guard (0...1).contains(clip.opacity) else {
                    throw ProjectValidationError.invalidOpacity(clip.id)
                }
                guard (0...2).contains(clip.audioVolume) else {
                    throw ProjectValidationError.invalidAudioVolume(clip.id)
                }
                let transform = clip.transform
                guard transform.scale > 0,
                      [transform.positionX, transform.positionY, transform.scale, transform.rotationDegrees,
                       transform.cropTop, transform.cropLeading, transform.cropBottom, transform.cropTrailing]
                        .allSatisfy(\.isFinite),
                      [transform.cropTop, transform.cropLeading, transform.cropBottom, transform.cropTrailing]
                        .allSatisfy({ (0...1).contains($0) }) else {
                    throw ProjectValidationError.invalidTransform(clip.id)
                }
                let fades = [clip.fades.videoIn, clip.fades.videoOut, clip.fades.audioIn, clip.fades.audioOut]
                guard fades.allSatisfy({ $0 >= .zero && $0 <= clip.duration }),
                      clip.fades.videoIn + clip.fades.videoOut <= clip.duration,
                      clip.fades.audioIn + clip.fades.audioOut <= clip.duration else {
                    throw ProjectValidationError.invalidFade(clip.id)
                }
                let color = clip.colorAdjustments
                guard [color.exposure, color.contrast, color.saturation, color.temperature, color.tint,
                       color.highlights, color.shadows, color.sharpen, color.vignette].allSatisfy(\.isFinite),
                      (-4...4).contains(color.exposure), (0...4).contains(color.contrast),
                      (0...4).contains(color.saturation), (-1...1).contains(color.temperature),
                      (-1...1).contains(color.tint), (-1...1).contains(color.highlights),
                      (-1...1).contains(color.shadows), (0...1).contains(color.sharpen),
                      (0...1).contains(color.vignette) else {
                    throw ProjectValidationError.invalidColorAdjustments(clip.id)
                }
                var effectIDs = Set<UUID>()
                for effect in clip.effects {
                    guard effectIDs.insert(effect.id).inserted else {
                        throw ProjectValidationError.duplicateIdentifier(effect.id)
                    }
                    guard effect.amount.isFinite, (0...1).contains(effect.amount) else {
                        throw ProjectValidationError.invalidEffect(clip.id)
                    }
                }
                for transition in [clip.transitionIn, clip.transitionOut].compactMap({ $0 }) {
                    guard transition.duration > .zero, transition.duration <= clip.duration else {
                        throw ProjectValidationError.invalidTransition(clip.id)
                    }
                }
                guard validKeyframes(clip.keyframes, duration: clip.duration) else {
                    throw ProjectValidationError.invalidKeyframe(clip.id)
                }
                switch clip.role {
                case .standard: break
                case .subtitle where clip.kind == .text: break
                case .voiceover where clip.kind == .audio: break
                default: throw ProjectValidationError.invalidRole(clip.id)
                }
                if index > 0 {
                    let previous = ordered[index - 1]
                    if previous.timeRange.intersects(clip.timeRange) {
                        throw ProjectValidationError.overlappingClips(previous.id, clip.id)
                    }
                }
            }
        }
    }

    private static func scaled(_ time: RationalTime, by factor: Double) -> RationalTime {
        RationalTime(seconds: time.seconds * factor, preferredTimescale: 60_000)
    }

    private static func validKeyframes(_ keyframes: ClipKeyframes, duration: RationalTime) -> Bool {
        for property in KeyframedProperty.allCases {
            let frames = keyframes[property]
            guard frames.allSatisfy({ $0.time >= .zero && $0.time <= duration && $0.value.isFinite }) else {
                return false
            }
            guard zip(frames, frames.dropFirst()).allSatisfy({ pair in pair.0.time < pair.1.time }) else { return false }
            switch property {
            case .opacity where !frames.allSatisfy({ (0...1).contains($0.value) }): return false
            case .volume where !frames.allSatisfy({ (0...2).contains($0.value) }): return false
            case .scale where !frames.allSatisfy({ $0.value > 0 }): return false
            default: break
            }
        }
        return true
    }
}

File: Windows/src/Cineleaf.Windows.Core/ProjectValidation.cs
namespace Cineleaf.Core;

public sealed class ProjectValidationException(string message) : Exception(message);

public static class ProjectValidator
{
    public static void Validate(CineleafProject project)
    {
        ArgumentNullException.ThrowIfNull(project);
        if (project.FormatVersion != CineleafProject.CurrentFormatVersion)
            throw new ProjectValidationException($"Unsupported project version {project.FormatVersion}.");
        if (project.Canvas.Width <= 0 || project.Canvas.Height <= 0)
            throw new ProjectValidationException("Canvas dimensions must be positive.");

        var identifiers = new HashSet<Guid> { project.Id };
        var assets = project.Assets.ToDictionary(asset => asset.Id);
        foreach (var asset in project.Assets)
        {
            if (!identifiers.Add(asset.Id)) throw new ProjectValidationException($"Duplicate identifier {asset.Id}.");
        }

        foreach (var marker in project.Timeline.Markers)
        {
            if (!identifiers.Add(marker.Id)) throw new ProjectValidationException($"Duplicate identifier {marker.Id}.");
            if (marker.Time < RationalTime.Zero) throw new ProjectValidationException($"Marker {marker.Id} has a negative time.");
        }

        foreach (var track in project.Timeline.Tracks)
        {
            if (!identifiers.Add(track.Id)) throw new ProjectValidationException($"Duplicate identifier {track.Id}.");
            var ordered = track.Clips.OrderBy(clip => clip.TimelineStart).ToList();
            for (var index = 0; index < ordered.Count; index++)
            {
                var clip = ordered[index];
                if (!identifiers.Add(clip.Id)) throw new ProjectValidationException($"Duplicate identifier {clip.Id}.");
                if (clip.TimelineStart < RationalTime.Zero || clip.SourceStart < RationalTime.Zero || clip.Duration <= RationalTime.Zero)
                    throw new ProjectValidationException($"Clip {clip.Id} has invalid time values.");
                if (index > 0 && ordered[index - 1].TimelineEnd > clip.TimelineStart)
                    throw new ProjectValidationException($"Clips {ordered[index - 1].Id} and {clip.Id} overlap.");
                if ((clip.Kind == ClipKind.Audio ? TrackKind.Audio : TrackKind.Video) != track.Kind)
                    throw new ProjectValidationException($"Clip {clip.Id} is on an incompatible track.");
                if (clip.Kind == ClipKind.Text)
                {
                    if (clip.TextStyle is null || clip.AssetId is not null)
                        throw new ProjectValidationException($"Text clip {clip.Id} is malformed.");
                }
                else if (clip.AssetId is not { } assetId || !assets.TryGetValue(assetId, out var asset))
                {
                    throw new ProjectValidationException($"Clip {clip.Id} references missing media.");
                }
                else if (clip.Kind != ClipKind.Image && asset.Metadata.Duration is { } sourceDuration &&
                         clip.SourceStart + RationalTime.FromSeconds(clip.Duration.Seconds * clip.PlaybackRate) > sourceDuration)
                {
                    throw new ProjectValidationException($"Clip {clip.Id} extends beyond its source media.");
                }
                if (!double.IsFinite(clip.PlaybackRate) || clip.PlaybackRate is < 0.25 or > 4)
                    throw new ProjectValidationException($"Clip {clip.Id} has an invalid playback rate.");
                if (!double.IsFinite(clip.Opacity) || clip.Opacity is < 0 or > 1)
                    throw new ProjectValidationException($"Clip {clip.Id} has invalid opacity.");
                if (!double.IsFinite(clip.AudioVolume) || clip.AudioVolume is < 0 or > 2)
                    throw new ProjectValidationException($"Clip {clip.Id} has invalid volume.");
                ValidateTransform(clip);
                ValidateFades(clip);
                ValidateAdvancedProperties(clip, identifiers);
            }
        }
    }

    private static void ValidateAdvancedProperties(TimelineClip clip, HashSet<Guid> identifiers)
    {
        var color = clip.ColorAdjustments;
        var colorValues = new[] { color.Exposure, color.Contrast, color.Saturation, color.Temperature, color.Tint,
            color.Highlights, color.Shadows, color.Sharpen, color.Vignette };
        if (colorValues.Any(value => !double.IsFinite(value)) || color.Exposure is < -4 or > 4 ||
            color.Contrast is < 0 or > 4 || color.Saturation is < 0 or > 4 ||
            color.Temperature is < -1 or > 1 || color.Tint is < -1 or > 1 ||
            color.Highlights is < -1 or > 1 || color.Shadows is < -1 or > 1 ||
            color.Sharpen is < 0 or > 1 || color.Vignette is < 0 or > 1)
            throw new ProjectValidationException($"Clip {clip.Id} has invalid color adjustments.");

        foreach (var effect in clip.Effects)
        {
            if (!identifiers.Add(effect.Id)) throw new ProjectValidationException($"Duplicate identifier {effect.Id}.");
            if (!double.IsFinite(effect.Amount) || effect.Amount is < 0 or > 1)
                throw new ProjectValidationException($"Clip {clip.Id} has an invalid effect.");
        }

        foreach (var transition in new[] { clip.TransitionIn, clip.TransitionOut }.OfType<ClipTransition>())
        {
            if (transition.Duration <= RationalTime.Zero || transition.Duration > clip.Duration)
                throw new ProjectValidationException($"Clip {clip.Id} has an invalid transition.");
        }

        ValidateKeyframes(clip, clip.Keyframes.PositionX, static _ => true);
        ValidateKeyframes(clip, clip.Keyframes.PositionY, static _ => true);
        ValidateKeyframes(clip, clip.Keyframes.RotationDegrees, static _ => true);
        ValidateKeyframes(clip, clip.Keyframes.Scale, static value => value > 0);
        ValidateKeyframes(clip, clip.Keyframes.Opacity, static value => value is >= 0 and <= 1);
        ValidateKeyframes(clip, clip.Keyframes.Volume, static value => value is >= 0 and <= 2);

        if ((clip.Role == ClipRole.Subtitle && clip.Kind != ClipKind.Text) ||
            (clip.Role == ClipRole.Voiceover && clip.Kind != ClipKind.Audio))
            throw new ProjectValidationException($"Clip {clip.Id} has an incompatible role.");
    }

    private static void ValidateKeyframes(TimelineClip clip, List<ScalarKeyframe> frames, Func<double, bool> valueIsValid)
    {
        for (var index = 0; index < frames.Count; index++)
        {
            var frame = frames[index];
            if (frame.Time < RationalTime.Zero || frame.Time > clip.Duration || !double.IsFinite(frame.Value) ||
                !valueIsValid(frame.Value) || (index > 0 && frames[index - 1].Time >= frame.Time))
                throw new ProjectValidationException($"Clip {clip.Id} has invalid keyframes.");
        }
    }

    private static void ValidateTransform(TimelineClip clip)
    {
        var transform = clip.Transform;
        var numbers = new[] { transform.PositionX, transform.PositionY, transform.Scale, transform.RotationDegrees,
            transform.CropTop, transform.CropLeading, transform.CropBottom, transform.CropTrailing };
        if (numbers.Any(value => !double.IsFinite(value)) || transform.Scale <= 0 ||
            new[] { transform.CropTop, transform.CropLeading, transform.CropBottom, transform.CropTrailing }
                .Any(value => value is < 0 or > 1))
            throw new ProjectValidationException($"Clip {clip.Id} has an invalid transform.");
    }

    private static void ValidateFades(TimelineClip clip)
    {
        var fades = new[] { clip.Fades.VideoIn, clip.Fades.VideoOut, clip.Fades.AudioIn, clip.Fades.AudioOut };
        if (fades.Any(fade => fade < RationalTime.Zero || fade > clip.Duration) ||
            clip.Fades.VideoIn + clip.Fades.VideoOut > clip.Duration ||
            clip.Fades.AudioIn + clip.Fades.AudioOut > clip.Duration)
            throw new ProjectValidationException($"Clip {clip.Id} has invalid fades.");
    }
}
